### PR TITLE
Bitcoin withdrawal service

### PR DESCRIPTION
**Pull request by dolgopolovwork**
Bitcoin transactions creation mechanism + new integration tests.

### CODE REVIEW

File: btc-withdrawal/src/main/kotlin/withdrawal/transaction/TransactionCreator.kt [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    fun createTransaction(
+        wallet: Wallet,
+        amount: Long,
+        destinationAddress: String,
+        confidenceLevel: Int
+    ): Result<Transaction, Exception> {
+        val transaction = Transaction(btcNetworkConfigProvider.getConfig())
+        return transactionHelper.getAvailableAddresses(wallet).flatMap { availableAddresses ->
+            transactionHelper.collectUnspents(availableAddresses, wallet, amount, confidenceLevel)
+        }.map { unspents ->
```
Comment (Alexey-N-Chernyshov) [inline btc-withdrawal/src/main/kotlin/withdrawal/transaction/TransactionCreator.kt]: Consider this:
```
            unspents.forEach { unspent -> transaction.addInput(unspent) }
            transactionHelper.addOutputs(transaction, unspents, destinationAddress, amount, changeAddress)
            transactionHelper.registerUnspents(unspents)
        }.map { transaction }
```

File: btc-withdrawal/src/main/kotlin/withdrawal/btc/config/BtcWithdrawalAppConfiguration.kt [hunk 2]
```diff
@@ -29,4 +33,26 @@ class BtcWithdrawalAppConfiguration {
         withdrawalConfig.iroha.port,
         withdrawalCredential
     )
+
+    @Bean
+    fun changeAddress(btcNetworkConfigProvider: BtcNetworkConfigProvider) =
```
[the remark this answers is not in the document]
Comment (grimadas) [reply btc-withdrawal/src/main/kotlin/withdrawal/btc/config/BtcWithdrawalAppConfiguration.kt]: Change address here is also a noun. Look up here https://en.bitcoin.it/wiki/Change

File: btc-withdrawal/src/main/kotlin/withdrawal/btc/config/BtcWithdrawalConfig.kt [hunk 3]
```diff
@@ -8,6 +9,14 @@ interface BtcWithdrawalConfig {
     val healthCheckPort: Int
     // Account that handles withdrawal events
     val withdrawalCredential: IrohaCredentialConfig
+    // Account that stores registered addresses
+    val registrationAccount: String
     // Iroha configurations
     val iroha: IrohaConfig
+    //Bitcoin configurations
+    val bitcoin: BitcoinConfig
+    // Base58 formatted Bitcoin address that stores changes
```
Comment (grimadas) [inline btc-withdrawal/src/main/kotlin/withdrawal/btc/config/BtcWithdrawalConfig.kt]: Change is uncountable

File: btc-withdrawal/src/main/kotlin/withdrawal/transaction/TransactionCreator.kt [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+class TransactionCreator(
+    @Autowired private val changeAddress: Address,
+    @Autowired private val btcNetworkConfigProvider: BtcNetworkConfigProvider,
+    @Autowired private val transactionHelper: TransactionHelper
+) {
+
+    /**
+     * Creates UNSIGNED Bitcoin transaction
+     * @param wallet - current wallet that will be used to fetch unspent transactions
+     * @param amount - amount of money to spend
```
Comment (grimadas) [inline btc-withdrawal/src/main/kotlin/withdrawal/transaction/TransactionCreator.kt]: Please write that amount is in satoshi

File: btc-withdrawal/src/main/kotlin/withdrawal/transaction/TransactionHelper.kt [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+) {
+
+    private val usedOutputs = HashSet<TransactionOutput>()
+
+    /**
+     * Adds outputs(destination and change addresses) to a given transaction
+     * @param transaction - current transaction
+     * @param unspents - used to compute change
+     * @param destinationAddress - receiver's base58 Bitcoin address
+     * @param amount - amount of money to spend(used to compute change)
```
Comment (grimadas) [inline btc-withdrawal/src/main/kotlin/withdrawal/transaction/TransactionHelper.kt]: in satoshi

File: btc-withdrawal/src/main/kotlin/withdrawal/transaction/TransactionHelper.kt [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+     * @param changeAddress - address that is used to store changes
```
Comment (grimadas) [inline btc-withdrawal/src/main/kotlin/withdrawal/transaction/TransactionHelper.kt]: *send change

File: btc-withdrawal/src/main/kotlin/withdrawal/btc/config/BtcWithdrawalAppConfiguration.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (dolgopolovwork) [reply btc-withdrawal/src/main/kotlin/withdrawal/btc/config/BtcWithdrawalAppConfiguration.kt]: "change" means "сдача" in this context

File: btc-withdrawal/src/main/kotlin/withdrawal/transaction/TransactionHelper.kt [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+            }
+            unspents
+        }
+    }
+
+    /**
+     * Registers given transaction outputs as "untouchable" to use in the future
+     * @param unspents - transaction outputs to register as "untouchable"
+     */
+    fun registerUnspents(unspents: List<TransactionOutput>) {
```
Comment (grimadas) [inline btc-withdrawal/src/main/kotlin/withdrawal/transaction/TransactionHelper.kt]: Where is it used?

Comment (dolgopolovwork) [reply btc-withdrawal/src/main/kotlin/withdrawal/transaction/TransactionHelper.kt]: It is used in createTransaction()

**Pull request merged**